ostd/src/mm/io.rs

```diff
@@ -672,3 +730,26 @@ impl<'a> From<&'a mut [u8]> for VmWriter<'a> {
         unsafe { Self::from_kernel_space(slice.as_mut_ptr(), slice.len()) }
     }
 }
+
+/// Asserts that reading or writing a `T` can be implemented with a single instruction.
+///
+/// We currently rely on this assertion to ensure that the memory operation created by
+/// `ptr::read_volatile` and `ptr::write_volatile` doesn't tear. However, the Rust documentation
+/// makes no such guarantee, and even the wording in the LLVM LangRef is ambiguous.
+///
+/// At this point, we can only _hope_ that this doesn't break in future versions of the Rust or
+/// LLVM compilers. However, this is unlikely to happen in practice, since the Linux kernel also
+/// uses "volatile" semantics to implement `READ_ONCE`/`WRITE_ONCE`.
+#[cfg(target_arch = "x86_64")]
+const fn assert_memory_access_no_tearing<T: Pod>() {
```

Comment:
Why not use `const_assert`?

```rust
const_assert! {
    let size = size_of::<T>();
    size == 0 || size == 1 || size == 2 || size == 4 || size == 8   
}
```

Reply:
I think I have tried this before. But it won't work due to a compiler error:
```
error[E0401]: can't use generic parameters from outer item
   --> ostd/src/mm/io.rs:748:30
    |
744 | const fn assert_memory_access_no_tearing<T: Pod>() {
    |                                          - type parameter from outer item
...
748 |         let size = size_of::<T>();
    |                              ^ use of generic parameter from outer item
    |
    = note: a `const` is a separate item from the item that contains it
```

Reply:
I see. It seems that we have encountered a limitation of the `static_assertion` crate.

There is another crate called `const-assert` that allows us to do this (an link to [Rust playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=82fc2ebb4e706b33507339de991b3918)).

```rust
pub struct VmReader {}

impl VmReader {
    pub fn read_once<T>(&self, offset: usize) -> T
    where
        Assert<{ is_atomic_type::<T>() }>: IsTrue
    {
        todo!()
    }
}

pub const fn is_atomic_type<T>() -> bool {
    let size = core::mem::size_of::<T>();
    size == 1 || size == 2 || size == 4 || size == 8    
}
```

No one seems to be using this crate. But [the crate is so small and obvious](https://github.com/Bajix/const-assert/blob/main/src/lib.rs) that it can hardly do anything wrong. This approach is more readable than our current approach.